File: Project1.scala
package scalation
package modeling


import scalation.mathstat._
import scala.math.{sqrt}
import scalation.mathstat.Scala2LaTeX._


def InSample (ox: MatrixD,  y: VectorD, ox_fname: Array[String], data_name: String): Unit =

    val n_q = 15
    val r_q = 0 until 15
    val colName = "Metric, Regression, Ridge, Lasso, Sqrt, log1p"
    val rowName = modeling.qoF_names.take(n_q)  

    val reg = new Regression (ox, y, ox_fname)
    val qof1 = reg.trainNtest ()()._2(r_q) // train and test on the same data (in-sample)

    val ox_centered = ox - ox.mean(0) // center the data by subtracting the mean of each column
    val y_centered  = y - y.mean         // center the target variable

    val ridge = new RidgeRegression (ox_centered, y_centered, ox_fname)
    val qof2 = ridge.trainNtest ()()._2(r_q) // train and test on the same data (in-sample)

    val lasso = new LassoRegression (ox, y, ox_fname)
    val qof3 = lasso.trainNtest ()()._2(r_q) // train and test on the same data (in-sample)

    val Sqrt = new TranRegression (ox, y, ox_fname, tran=sqrt, itran=sq)
    val qof4 = Sqrt.trainNtest ()()._2(r_q) // train and test on the same data (in-sample)

    val Log1p = new TranRegression (ox, y, ox_fname)
    val qof5 = Log1p.trainNtest ()()._2(r_q) // train and test on the same data (in-sample)

    
    val caption = data_name + " In-Sample QoF Comparison"
    val name    = "tab:" + caption
    val qof_matrix = MatrixD (qof1, qof2, qof3, qof4, qof5).transpose // transpose to get metrics as rows and models as columns

    val latex   = make_doc (make_table (caption, name, qof_matrix, colName, rowName))
    println (latex)

end InSample

@main def P1_AutoMPG_IS (): Unit =
    import Example_AutoMPG._
    InSample (ox, y, ox_fname, "Auto MPG")
end P1_AutoMPG_IS

@main def P1_Housing_IS (): Unit =
    val ox_fname = Array ("intercept", "Square_Footage", "Num_Bedrooms", "Num_Bathrooms", "Year_Built", "Lot_Size", "Garage_Size", "Neighborhood_Quality")
    val xy = MatrixD.load ("house_price_regression_dataset.csv", 1, sp=',')      // load the dataset, skipping the header row
    val x = xy(?, 0 until 6)                                  // get the first 6 columns as the feature matrix
    val y = xy(?, 7)                                        // get the 7th column as the response vector
    val _1     = VectorD.one (xy.dim)                                  // vector of all ones    
    val ox     = _1 +^: x                                              // prepend a column of all ones to x
    InSample (ox, y, ox_fname, "Housing Prices")
end P1_Housing_IS

@main def P1_Insurance_IS (): Unit =
    val ox_fname = Array ("intercept", "age", "bmi", "children", "sex_male", "smoker_yes", "region_northwest", "region_southeast", "region_southwest")
    val oxy = MatrixD.load ("insurance_cat2num.csv", 1, sp=',')      // load the dataset, skipping the header row
    val ox = oxy(?, 0 until 8)                                  // get the first 8 columns as the feature matrix
    val y = oxy(?, 9)                                        // get the 9th column as the response vector
    InSample(ox, y, ox_fname, "Insurance Charges")
end P1_Insurance_IS

def OutOfSample (ox: MatrixD,  y: VectorD, ox_fname: Array[String], data_name: String, testp: Double): Unit =

    val n_q = 15
    val r_q = 0 until 15
    val colName = "Metric, Regression, Ridge, Lasso, Sqrt, log1p"
    val rowName = modeling.qoF_names.take(n_q)

    val permGen = scalation.mathstat.TnT_Split.makePermGen (ox.dim)             // make a permutation generator
    val n_test = (ox.dim * testp).toInt                                           // (1-testp)% training, testp% testing
    val idx = scalation.mathstat.TnT_Split.testIndices(permGen, n_test)         // get test indices for 80-20 split
    val (ox_test, ox_train) = TnT_Split (ox, idx)                               // TnT split the dataset ox (row split)
    val yy = MatrixD.fromVector (y)                                // turn the m-vector y into an m-by-1 matrix 
    val (yy_test, yy_train) = TnT_Split (yy, idx)                               // TnT split the response vector y (row split)
    val y_train = yy_train.col(0)                                                     // get the test response vector from the test response matrix
    val y_test = yy_test.col(0)                                                       // get the test response vector from the test response matrix
    

    val reg = new Regression (ox, y, ox_fname)
    val qof1 = reg.trainNtest (ox_train, y_train)(ox_test, y_test)._2(r_q) // train and test on the same data (in-sample)

    val ox_centered = ox - ox_train.mean(0) // center the data by subtracting the mean of each column
    val y_centered  = y - y_train.mean         // center the target variable
    val ox_train_centered = ox_train - ox_train.mean(0) // center the data by subtracting the mean of each column
    val y_train_centered  = y_train - y_train.mean         // center the target variable
    val ox_test_centered = ox_test - ox_train.mean(0) // center the data by subtracting the mean of each column
    val y_test_centered  = y_test - y_train.mean         // center the target variable

    val ridge = new RidgeRegression (ox_centered, y_centered, ox_fname)
    val qof2 = ridge.trainNtest (ox_train_centered, y_train_centered)(ox_test_centered, y_test_centered)._2(r_q) // train and test on the same data (in-sample)

    val lasso = new LassoRegression (ox, y, ox_fname)
    val qof3 = lasso.trainNtest (ox_train, y_train)(ox_test, y_test)._2(r_q) // train and test on the same data (in-sample)

    val Sqrt = new TranRegression (ox, y, ox_fname, tran=sqrt, itran=sq)
    // val qof4 = Sqrt.trainNtest (ox_train, y_train)(ox_test, y_test)._2(r_q) // train and test on the same data (in-sample)
    val (sqrtyp, qof4_full) = Sqrt.validate ()() // train on the training data and validate on the test data (validation)
    val qof4 = qof4_full(r_q) // get the QoF metrics for the core metrics only (r_q)
    val (y_test_ord_1, sqrtyp_ord) = orderByY (y_test, sqrtyp) // order the test response and the sqrt predictions by the test response for better visualization
    new Plot (null, y_test_ord_1, sqrtyp_ord, s"Plot ${Sqrt.modelName} predictions: yy black/actual vs. yp red/predicted", lines = true)

    val Log1p = new TranRegression (ox, y, ox_fname)
    // val qof5 = Log1p.trainNtest (ox_train, y_train)(ox_test, y_test)._2(r_q) // train and test on the same data (in-sample)
    val (log1pyp, qof5_full) = Log1p.validate ()() // train on the training data and validate on the test data (validation)
    val qof5 = qof5_full(r_q) // get the QoF metrics for the core metrics only (r_q)
    val (y_test_ord_2, log1yp_ord) = orderByY (y_test, log1pyp) // order the test response and the log1p predictions by the test response for better visualization
    new Plot (null, y_test_ord_2, log1yp_ord, s"Plot ${Log1p.modelName} predictions: yy black/actual vs. yp red/predicted", lines = true)
    
    val caption = data_name + " Out-of-Sample QoF Comparison"
    val name    = "tab:" + caption
    val qof_matrix = MatrixD (qof1, qof2, qof3, qof4, qof5).transpose // transpose to get metrics as rows and models as columns

    val latex   = make_doc (make_table (caption, name, qof_matrix, colName, rowName))
    println (latex)

end OutOfSample

@main def P1_AutoMPG_OOS (): Unit =
    import Example_AutoMPG._
    OutOfSample (ox, y, ox_fname, "Auto MPG", 0.2)
end P1_AutoMPG_OOS

@main def P1_Housing_OOS (): Unit =
    val ox_fname = Array ("intercept", "Square_Footage", "Num_Bedrooms", "Num_Bathrooms", "Year_Built", "Lot_Size", "Garage_Size", "Neighborhood_Quality")
    val xy = MatrixD.load ("house_price_regression_dataset.csv", 1, sp=',')      // load the dataset, skipping the header row
    val x = xy(?, 0 until 6)                                  // get the first 6 columns as the feature matrix
    val y = xy(?, 7)                                        // get the 7th column as the response vector
    val _1     = VectorD.one (xy.dim)                                  // vector of all ones    
    val ox     = _1 +^: x                                              // prepend a column of all ones to x
    OutOfSample (ox, y, ox_fname, "Housing Prices", 0.2)
end P1_Housing_OOS

@main def P1_Insurance_OOS (): Unit =
    val ox_fname = Array ("intercept", "age", "bmi", "children", "sex_male", "smoker_yes", "region_northwest", "region_southeast", "region_southwest")
    val oxy = MatrixD.load ("insurance_cat2num.csv", 1, sp=',')      // load the dataset, skipping the header row
    val ox = oxy(?, 0 until 8)                                  // get the first 8 columns as the feature matrix
    val y = oxy(?, 9)                                        // get the 9th column as the response vector
    OutOfSample(ox, y, ox_fname, "Insurance Charges", 0.2)
end P1_Insurance_OOS

File: Project1.py
import statsmodels.api as sm
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split


def getQoF(y_actual, y_pred, k, mod):
    # Get the number of observations (m)
    m = len(y_actual)  # Number of observations in the original dataset

    # Calculate SSE (Sum of Squared Errors)
    try:
        sse = mod.ssr
    except:
        sse = np.sum((y_actual - y_pred)**2)

    # Calculate SST (Total Sum of Squares)
    try:
        sst = mod.centered_tss
    except:
        sst= np.sum((y_actual - np.mean(y_actual))**2)

    # Compute R-squared
    try:
        rsquared = mod.rsquared
    except:
        rsquared = 1 - (sse / sst)

    # Calculate Adjusted R-squared
    try:
        rsquared_adj = mod.rsquared_adj
    except:
        rsquared_adj = 1 - (1 - rsquared) * (m - 1) / (m - k)

    # Standard Deviation of the Errors (SDE)
    try:
        sde = np.sqrt(mod.mse_resid)
    except:
        sde = np.sqrt(sse / (m - k))

    # Calculate Mean Squared Error
    try:
        mse0 = mod.mse_resid
    except:
        mse0 = sse / m

    # Root Mean Squared Error
    try:
        rmse = np.sqrt(mod.mse_resid)
    except:
        rmse = np.sqrt(mse0)

    # Mean Absolute Error
    try:
        mae = mod.mae
    except:
        mae = np.mean(np.abs(y_actual - y_pred))

    # Symmetric Mean Absolute Percentage Error (SMAPE)
    try:
        smape = mod.smape
    except:
        smape = np.mean(2 * np.abs(y_actual - y_pred) / (np.abs(y_actual) + np.abs(y_pred))) * 100

    # Calculate F-statistic
    try:
        f_stat = mod.fvalue
    except:
        f_stat = ((sst - sse) / k) / (sse / (m - k))

    # Get the degree of freedom for the model
    try:
        dfr = mod.df_model
    except:
        dfr = k

    # Get the degree of freedom for the residuals
    try:
        df = mod.df_resid
    except:
        df = m - k

    # Calculate AIC
    try:
        aic = mod.aic
    except:
        aic = m * np.log(mse0) + 2 * k

    # Calculate BIC
    try:
        bic = mod.bic
    except:
        bic = m * np.log(mse0) + k * np.log(m)

    QoF = list(range(15))

    QoF[0] = rsquared
    QoF[1] = rsquared_adj
    QoF[2] = sst
    QoF[3] = sse
    QoF[4] = sde
    QoF[5] = mse0
    QoF[6] = rmse
    QoF[7] = mae
    QoF[8] = smape
    QoF[9] = m
    QoF[10] = dfr
    QoF[11] = df
    QoF[12] = f_stat
    QoF[13] = aic
    QoF[14] = bic

    return QoF




def save_sorted_plot(y_actual, y_pred, data_name, model_name, validate=False):
    # Create a DataFrame with actual and predicted values
    df = pd.DataFrame({'Actual': y_actual, 'Predicted': y_pred})
    # Sort the DataFrame by actual values
    df_sorted = df.sort_values(by='Actual')
    x_end = df_sorted.shape[0]  # Number of observations
    x = np.arange(0, x_end)  # X-axis values from 1 to number of observations
    y_start = min(df_sorted['Actual'].min(), df_sorted['Predicted'].min())
    y_end = max(df_sorted['Actual'].max(), df_sorted['Predicted'].max())
    y_dist = y_end - y_start
    y_start = y_start - 0.1 * y_dist
    y_end = y_end + 0.1 * y_dist
    x_end_2 = int(1.1 * x_end) if x_end > 0 else int(0.9 * x_end)
    x_start = int(-0.1 * x_end)
    # Plot the sorted data
    plt.figure(figsize=(10, 6))
    plt.plot(x, df_sorted['Predicted'], label='Predicted Values', color='red')
    plt.plot(x, df_sorted['Actual'], color='black', label='Actual Values')
    plt.title(f'{data_name} Linear Regression, {"80-20 Split" if validate else "In-Sample"}: yy black/actual vs. yp red/predicted')
    plt.legend()
    plt.ylim(y_start, y_end)
    plt.xlim(x_start, x_end_2)
    plt.savefig(f'statsmodels_{model_name}_{"80_20" if validate else "In_Sample"}.png')




def InSample(X, y, data_name):
    k = X.shape[1] - 1  # Number of predictors (excluding the intercept)

    reg = sm.OLS(y, X).fit()
    regyp = reg.predict(X)
    reg_QoF = getQoF(y, regyp, k, reg)

    ridge = sm.OLS(y, X).fit_regularized(method='elastic_net', L1_wt=0, alpha=1.0)
    ridgeyp = ridge.predict(X)
    ridge_QoF = getQoF(y, ridgeyp, k, ridge)

    lasso = sm.OLS(y, X).fit_regularized(method='elastic_net', L1_wt=1, alpha=1.0)
    lassoyp = lasso.predict(X)
    lasso_QoF = getQoF(y, lassoyp, k, lasso)

    Sqrtf = sm.OLS(np.sqrt(y), X).fit()
    Sqrtfyp_tran = Sqrtf.predict(X)
    Sqrtfyp = Sqrtfyp_tran ** 2
    Sqrtf_QoF = getQoF(y, Sqrtfyp, k, Sqrtf)

    log1p = sm.OLS(np.log1p(y), X).fit()
    log1pyp_tran = log1p.predict(X)
    log1pyp = np.expm1(log1pyp_tran)
    log1p_QoF = getQoF(y, log1pyp, k, log1p)

    print("\\begin{table}[h]")
    print("\\centering")
    print(f"\\caption{{Statsmodels - {data_name} In-Sample QoF Comparison}}")
    print(f"\\label{{tab:Statsmodels - {data_name} In-Sample QoF Comparison}}")
    print("\\begin{tabular}{|c|c|c|c|c|c|}\\hline")
    print("Metric & Regression & Ridge & Lasso & Sqrt & Log1p \\\\ \\hline \\hline")
    print(f"rSq & {reg_QoF[0]:.4f} & {ridge_QoF[0]:.4f} & {lasso_QoF[0]:.4f} & {Sqrtf_QoF[0]:.4f} & {log1p_QoF[0]:.4f} \\\\ \\hline")
    print(f"rSqBar & {reg_QoF[1]:.4f} & {ridge_QoF[1]:.4f} & {lasso_QoF[1]:.4f} & {Sqrtf_QoF[1]:.4f} & {log1p_QoF[1]:.4f} \\\\ \\hline")
    print(f"sst & {reg_QoF[2]:.4f} & {ridge_QoF[2]:.4f} & {lasso_QoF[2]:.4f} & {Sqrtf_QoF[2]:.4f} & {log1p_QoF[2]:.4f} \\\\ \\hline")
    print(f"sse & {reg_QoF[3]:.4f} & {ridge_QoF[3]:.4f} & {lasso_QoF[3]:.4f} & {Sqrtf_QoF[3]:.4f} & {log1p_QoF[3]:.4f} \\\\ \\hline")
    print(f"sde & {reg_QoF[4]:.4f} & {ridge_QoF[4]:.4f} & {lasso_QoF[4]:.4f} & {Sqrtf_QoF[4]:.4f} & {log1p_QoF[4]:.4f} \\\\ \\hline")
    print(f"mse0 & {reg_QoF[5]:.4f} & {ridge_QoF[5]:.4f} & {lasso_QoF[5]:.4f} & {Sqrtf_QoF[5]:.4f} & {log1p_QoF[5]:.4f} \\\\ \\hline")
    print(f"rmse & {reg_QoF[6]:.4f} & {ridge_QoF[6]:.4f} & {lasso_QoF[6]:.4f} & {Sqrtf_QoF[6]:.4f} & {log1p_QoF[6]:.4f} \\\\ \\hline")
    print(f"mae & {reg_QoF[7]:.4f} & {ridge_QoF[7]:.4f} & {lasso_QoF[7]:.4f} & {Sqrtf_QoF[7]:.4f} & {log1p_QoF[7]:.4f} \\\\ \\hline")
    print(f"smape & {reg_QoF[8]:.4f} & {ridge_QoF[8]:.4f} & {lasso_QoF[8]:.4f} & {Sqrtf_QoF[8]:.4f} & {log1p_QoF[8]:.4f} \\\\ \\hline")
    print(f"m & {reg_QoF[9]:.4f} & {ridge_QoF[9]:.4f} & {lasso_QoF[9]:.4f} & {Sqrtf_QoF[9]:.4f} & {log1p_QoF[9]:.4f} \\\\ \\hline")
    print(f"dfr & {reg_QoF[10]:.4f} & {ridge_QoF[10]:.4f} & {lasso_QoF[10]:.4f} & {Sqrtf_QoF[10]:.4f} & {log1p_QoF[10]:.4f}\\\\ \\hline")
    print(f"df & {reg_QoF[11]:.4f} & {ridge_QoF[11]:.4f} & {lasso_QoF[11]:.4f} & {Sqrtf_QoF[11]:.4f} & {log1p_QoF[11]:.4f}\\\\ \\hline")
    print(f"fStat & {reg_QoF[12]:.4f} & {ridge_QoF[12]:.4f} & {lasso_QoF[12]:.4f} & {Sqrtf_QoF[12]:.4f} &{log1p_QoF[12]:.4f}\\\\ \\hline")
    print(f"aic &{reg_QoF[13]:.4f} &{ridge_QoF[13]:.4f} & {lasso_QoF[13]:.4f} & {Sqrtf_QoF[13]:.4f} & {log1p_QoF[13]:.4f}\\\\ \\hline")
    print(f"bic & {reg_QoF[14]:.4f} & {ridge_QoF[14]:.4f} & {lasso_QoF[14]:.4f} & {Sqrtf_QoF[14]:.4f} & {log1p_QoF[14]:.4f}\\\\ \\hline")
    print("\\end{tabular}")
    print("\\end{table}")

    return (regyp, ridgeyp, lassoyp, Sqrtfyp, log1pyp)



def P1_AutoMPG_IS():
    Auto_MPG = pd.read_csv('auto_mpg_cleaned.csv')
    Auto_MPG = Auto_MPG.dropna()
    Auto_MPG = Auto_MPG.drop('origin', axis=1)
    X = Auto_MPG[['displacement', 'cylinders', 'horsepower', 'weight', 'acceleration', 'model_year']]
    y = Auto_MPG['mpg']
    X = sm.add_constant(X)
    yp = InSample(X, y, "Auto MPG")
    save_sorted_plot(y, yp[0], "Auto MPG", "reg", validate=False)
    save_sorted_plot(y, yp[1], "Auto MPG", "ridge", validate=False)
    save_sorted_plot(y, yp[2], "Auto MPG", "lasso", validate=False)
    save_sorted_plot(y, yp[3], "Auto MPG", "sqrt", validate=False)
    save_sorted_plot(y, yp[4], "Auto MPG", "log1p", validate=False)



def P1_Housing_IS():
    House_Price = pd.read_csv('house_price_regression_dataset.csv')
    X = House_Price[['Square_Footage', 'Num_Bedrooms', 'Num_Bathrooms', 'Year_Built', 'Lot_Size', 'Garage_Size', 'Neighborhood_Quality']]
    y = House_Price['House_Price']
    X = sm.add_constant(X)
    yp = InSample(X, y, "House Price")
    save_sorted_plot(y, yp[0], "House Price", "reg", validate=False)
    save_sorted_plot(y, yp[1], "House Price", "ridge", validate=False)
    save_sorted_plot(y, yp[2], "House Price", "lasso", validate=False)
    save_sorted_plot(y, yp[3], "House Price", "sqrt", validate=False)
    save_sorted_plot(y, yp[4], "House Price", "log1p", validate=False)



def P1_Insuarance_IS():
    Insurance_Charges = pd.read_csv('insurance_cat2num.csv')
    X = Insurance_Charges[['intercept', 'age', 'bmi', 'children', 'sex_male', 'smoker_yes', 'region_northwest', 'region_southeast', 'region_southwest']]
    y = Insurance_Charges['charges']
    yp = InSample(X, y, "Insurance Charges")
    save_sorted_plot(y, yp[0], "Insurance Charges", "reg", validate=False)
    save_sorted_plot(y, yp[1], "Insurance Charges", "ridge", validate=False)
    save_sorted_plot(y, yp[2], "Insurance Charges", "lasso", validate=False)
    save_sorted_plot(y, yp[3], "Insurance Charges", "sqrt", validate=False)
    save_sorted_plot(y, yp[4], "Insurance Charges", "log1p", validate=False)




# P1_AutoMPG_IS()
# P1_Housing_IS()
# P1_Insuarance_IS()




def OutOfSample(X, y, data_name, testp=0.2, random_state=42):
    # Split the dataset into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=testp, random_state=random_state)

    k = X.shape[1] - 1  # Number of predictors (excluding the intercept)

    reg = sm.OLS(y_train, X_train).fit()
    regyp = reg.predict(X_test)
    reg_QoF = getQoF(y_test, regyp, k, None)

    ridge = sm.OLS(y_train, X_train).fit_regularized(method='elastic_net', L1_wt=0, alpha=1.0)
    ridgeyp = ridge.predict(X_test)
    ridge_QoF = getQoF(y_test, ridgeyp, k, None)

    lasso = sm.OLS(y_train, X_train).fit_regularized(method='elastic_net', L1_wt=1, alpha=1.0)
    lassoyp = lasso.predict(X_test)
    lasso_QoF = getQoF(y_test, lassoyp, k, None)

    Sqrtf = sm.OLS(np.sqrt(y_train), X_train).fit()
    Sqrtfyp_tran = Sqrtf.predict(X_test)
    Sqrtfyp = Sqrtfyp_tran ** 2
    Sqrtf_QoF = getQoF(y_test, Sqrtfyp, k, None)

    log1p = sm.OLS(np.log1p(y_train), X_train).fit()
    log1pyp_tran = log1p.predict(X_test)
    log1pyp = np.expm1(log1pyp_tran)
    log1p_QoF = getQoF(y_test, log1pyp, k, None)

    print("\\begin{table}[h]")
    print("\\centering")
    print(f"\\caption{{Statsmodels - {data_name} Out-of-Sample QoF Comparison}}")
    print(f"\\label{{tab:Statsmodels - {data_name} Out-of-Sample QoF Comparison}}")
    print("\\begin{tabular}{|c|c|c|c|c|c|}\\hline")
    print("Metric & Regression & Ridge & Lasso & Sqrt & Log1p \\\\ \\hline \\hline")
    print(f"rSq & {reg_QoF[0]:.4f} & {ridge_QoF[0]:.4f} & {lasso_QoF[0]:.4f} & {Sqrtf_QoF[0]:.4f} & {log1p_QoF[0]:.4f} \\\\ \\hline")
    print(f"rSqBar & {reg_QoF[1]:.4f} & {ridge_QoF[1]:.4f} & {lasso_QoF[1]:.4f} & {Sqrtf_QoF[1]:.4f} & {log1p_QoF[1]:.4f} \\\\ \\hline")
    print(f"sst & {reg_QoF[2]:.4f} & {ridge_QoF[2]:.4f} & {lasso_QoF[2]:.4f} & {Sqrtf_QoF[2]:.4f} & {log1p_QoF[2]:.4f} \\\\ \\hline")
    print(f"sse & {reg_QoF[3]:.4f} & {ridge_QoF[3]:.4f} & {lasso_QoF[3]:.4f} & {Sqrtf_QoF[3]:.4f} & {log1p_QoF[3]:.4f} \\\\ \\hline")
    print(f"sde & {reg_QoF[4]:.4f} & {ridge_QoF[4]:.4f} & {lasso_QoF[4]:.4f} & {Sqrtf_QoF[4]:.4f} & {log1p_QoF[4]:.4f} \\\\ \\hline")
    print(f"mse0 & {reg_QoF[5]:.4f} & {ridge_QoF[5]:.4f} & {lasso_QoF[5]:.4f} & {Sqrtf_QoF[5]:.4f} & {log1p_QoF[5]:.4f} \\\\ \\hline")
    print(f"rmse & {reg_QoF[6]:.4f} & {ridge_QoF[6]:.4f} & {lasso_QoF[6]:.4f} & {Sqrtf_QoF[6]:.4f} & {log1p_QoF[6]:.4f} \\\\ \\hline")
    print(f"mae & {reg_QoF[7]:.4f} & {ridge_QoF[7]:.4f} & {lasso_QoF[7]:.4f} & {Sqrtf_QoF[7]:.4f} & {log1p_QoF[7]:.4f} \\\\ \\hline")
    print(f"smape & {reg_QoF[8]:.4f} & {ridge_QoF[8]:.4f} & {lasso_QoF[8]:.4f} & {Sqrtf_QoF[8]:.4f} & {log1p_QoF[8]:.4f} \\\\ \\hline")
    print(f"m & {reg_QoF[9]:.4f} & {ridge_QoF[9]:.4f} & {lasso_QoF[9]:.4f} & {Sqrtf_QoF[9]:.4f} & {log1p_QoF[9]:.4f} \\\\ \\hline")
    print(f"dfr & {reg_QoF[10]:.4f} & {ridge_QoF[10]:.4f} & {lasso_QoF[10]:.4f} & {Sqrtf_QoF[10]:.4f} & {log1p_QoF[10]:.4f}\\\\ \\hline")
    print(f"df & {reg_QoF[11]:.4f} & {ridge_QoF[11]:.4f} & {lasso_QoF[11]:.4f} & {Sqrtf_QoF[11]:.4f} & {log1p_QoF[11]:.4f}\\\\ \\hline")
    print(f"fStat & {reg_QoF[12]:.4f} & {ridge_QoF[12]:.4f} & {lasso_QoF[12]:.4f} & {Sqrtf_QoF[12]:.4f} &{log1p_QoF[12]:.4f}\\\\ \\hline")
    print(f"aic &{reg_QoF[13]:.4f} &{ridge_QoF[13]:.4f} & {lasso_QoF[13]:.4f} & {Sqrtf_QoF[13]:.4f} & {log1p_QoF[13]:.4f}\\\\ \\hline")
    print(f"bic & {reg_QoF[14]:.4f} & {ridge_QoF[14]:.4f} & {lasso_QoF[14]:.4f} & {Sqrtf_QoF[14]:.4f} & {log1p_QoF[14]:.4f}\\\\ \\hline")
    print("\\end{tabular}")
    print("\\end{table}")

    return (regyp, ridgeyp, lassoyp, Sqrtfyp, log1pyp, y_test)




def P1_AutoMPG_OOS():
    Auto_MPG = pd.read_csv('auto_mpg_cleaned.csv')
    Auto_MPG = Auto_MPG.dropna()
    Auto_MPG = Auto_MPG.drop('origin', axis=1)
    X = Auto_MPG[['displacement', 'cylinders', 'horsepower', 'weight', 'acceleration', 'model_year']]
    y = Auto_MPG['mpg']
    X = sm.add_constant(X)
    yp = OutOfSample(X, y, "Auto MPG", testp=0.2, random_state=42)
    save_sorted_plot(yp[5], yp[0], "Auto MPG", "reg", validate=True)
    save_sorted_plot(yp[5], yp[1], "Auto MPG", "ridge", validate=True)
    save_sorted_plot(yp[5], yp[2], "Auto MPG", "lasso", validate=True)
    save_sorted_plot(yp[5], yp[3], "Auto MPG", "sqrt", validate=True)
    save_sorted_plot(yp[5], yp[4], "Auto MPG", "log1p", validate=True)



def P1_Housing_OOS():
    House_Price = pd.read_csv('house_price_regression_dataset.csv')
    X = House_Price[['Square_Footage', 'Num_Bedrooms', 'Num_Bathrooms', 'Year_Built', 'Lot_Size', 'Garage_Size', 'Neighborhood_Quality']]
    y = House_Price['House_Price']
    X = sm.add_constant(X)
    yp = OutOfSample(X, y, "House Price", testp=0.2, random_state=42)
    save_sorted_plot(yp[5], yp[0], "House Price", "reg", validate=True)
    save_sorted_plot(yp[5], yp[1], "House Price", "ridge", validate=True)
    save_sorted_plot(yp[5], yp[2], "House Price", "lasso", validate=True)
    save_sorted_plot(yp[5], yp[3], "House Price", "sqrt", validate=True)
    save_sorted_plot(yp[5], yp[4], "House Price", "log1p", validate=True)



def P1_Insuarance_OOS():
    Insurance_Charges = pd.read_csv('insurance_cat2num.csv')
    X = Insurance_Charges[['intercept', 'age', 'bmi', 'children', 'sex_male', 'smoker_yes', 'region_northwest', 'region_southeast', 'region_southwest']]
    y = Insurance_Charges['charges']
    yp = OutOfSample(X, y, "Insurance Charges", testp=0.2, random_state=42)
    save_sorted_plot(yp[5], yp[0], "Insurance Charges", "reg", validate=True)
    save_sorted_plot(yp[5], yp[1], "Insurance Charges", "ridge", validate=True)
    save_sorted_plot(yp[5], yp[2], "Insurance Charges", "lasso", validate=True)
    save_sorted_plot(yp[5], yp[3], "Insurance Charges", "sqrt", validate=True)
    save_sorted_plot(yp[5], yp[4], "Insurance Charges", "log1p", validate=True)




# P1_AutoMPG_OOS()
# P1_Housing_OOS()
# P1_Insuarance_OOS()